File: dotnet-adapters-shortest-paths-impl-yanqi/src/main/java/com/programmerare/shortestpaths/adapter/yanqi/generics/PathFinderFactoryYanQiGenerics.java
package com.programmerare.shortestpaths.adapter.yanqi.generics;

import com.programmerare.shortestpaths.core.api.Vertex;
import com.programmerare.shortestpaths.core.api.Weight;
import com.programmerare.shortestpaths.core.api.generics.EdgeGenerics;
import com.programmerare.shortestpaths.core.api.generics.GraphGenerics;
import com.programmerare.shortestpaths.core.api.generics.PathFinderFactoryGenerics;
import com.programmerare.shortestpaths.core.api.generics.PathFinderGenerics;
import com.programmerare.shortestpaths.core.api.generics.PathGenerics;
import com.programmerare.shortestpaths.core.impl.generics.PathFinderFactoryGenericsBase;

/**
 * @author Tomas Johansson
 */
public class PathFinderFactoryYanQiGenerics<F extends PathFinderGenerics<P, E,V,W> , P extends PathGenerics<E, V, W> ,  E extends EdgeGenerics<V, W> , V extends Vertex , W extends Weight> 
	extends PathFinderFactoryGenericsBase<F,P,E,V,W> 
	implements PathFinderFactoryGenerics<F,P,E,V,W> 
{
	public F createPathFinder(
		final GraphGenerics<E, V, W> graph 
	) {
		// TODO: try to get rid of the casting below ( warning: "Type safety: Unchecked cast from PathFinderYanQi<P,E,V,W> to F" )
		return (F) new PathFinderYanQiGenerics<P, E, V, W>(
			graph 
		);
	}	
}

File: dotnet-adapters-shortest-paths-impl-yanqi/src/com.programmerare.shortestpaths.adapter.yanqi/generics/PathFinderFactoryYanQiGenerics.cs
/*
* Copyright (c) Tomas Johansson , http://www.programmerare.com
* Regarding the license (Apache), please find more information 
* in the file "LICENSE_NOTICE.txt" in the project root directory 
* and also license information at this URL:
* https://github.com/TomasJohansson/adapters-shortest-paths-dotnet/
*/

using Programmerare.ShortestPaths.Core.Api;
using Programmerare.ShortestPaths.Core.Api.Generics;
using Programmerare.ShortestPaths.Core.Impl.Generics;

namespace Programmerare.ShortestPaths.Adapter.YanQi.Generics
{
    public class PathFinderFactoryYanQiGenerics<F, P, E, V, W>
        : PathFinderFactoryGenericsBase<F, P, E, V, W>
        , PathFinderFactoryGenerics<F, P, E, V, W>
        where F : PathFinderGenerics<P, E, V, W>
        where P : PathGenerics<E, V, W>
        where E : EdgeGenerics<V, W>
        where V : Vertex
        where W : Weight
    {
        public override F CreatePathFinder(GraphGenerics<E, V, W> graph)
        {
            PathFinderGenerics<P, E, V, W> pathFinder =  new PathFinderYanQiGenerics<P, E, V, W>(graph);
            // TODO: try to get rid of the casting below ( warning: "Type safety: Unchecked cast from PathFinderYanQi<P,E,V,W> to F" )
            return (F) pathFinder;
        }
    }
}